src/hotspot/share/runtime/sharedRuntime.cpp

```diff
@@ -2637,150 +2734,178 @@ AdapterHandlerEntry* AdapterHandlerLibrary::new_entry(AdapterFingerPrint* finger
+  bool contains_all_checks = StubRoutines::code2() != NULL;
+
+  VMRegPair stack_regs[16];
+  VMRegPair* regs = (total_args_passed <= 16) ? stack_regs : NEW_RESOURCE_ARRAY(VMRegPair, total_args_passed);
+
+  // Get a description of the compiled java calling convention and the largest used (VMReg) stack slot usage
+  int comp_args_on_stack = SharedRuntime::java_calling_convention(sig_bt, regs, total_args_passed);
+  BufferBlob* buf = buffer_blob(); // the temporary code buffer in CodeCache
+  CodeBuffer buffer(buf);
+  short buffer_locs[20];
+  buffer.insts()->initialize_shared_locs((relocInfo*)buffer_locs,
+                                          sizeof(buffer_locs)/sizeof(relocInfo));
+
+  // Make a C heap allocated version of the fingerprint to store in the adapter
+  AdapterFingerPrint* fingerprint = new AdapterFingerPrint(total_args_passed, sig_bt);
+  MacroAssembler _masm(&buffer);
+  AdapterHandlerEntry* entry = SharedRuntime::generate_i2c2i_adapters(&_masm,
+                                                total_args_passed,
+                                                comp_args_on_stack,
+                                                sig_bt,
+                                                regs,
+                                                fingerprint);
 
-      MacroAssembler _masm(&buffer);
-      entry = SharedRuntime::generate_i2c2i_adapters(&_masm,
-                                                     total_args_passed,
-                                                     comp_args_on_stack,
-                                                     sig_bt,
-                                                     regs,
-                                                     fingerprint);
 #ifdef ASSERT
-      if (VerifyAdapterSharing) {
-        if (shared_entry != NULL) {
-          assert(shared_entry->compare_code(buf->code_begin(), buffer.insts_size()), "code must match");
-          // Release the one just created and return the original
-          _adapters->free_entry(entry);
-          return shared_entry;
-        } else  {
-          entry->save_code(buf->code_begin(), buffer.insts_size());
-        }
-      }
+  if (VerifyAdapterSharing) {
+    entry->save_code(buf->code_begin(), buffer.insts_size());
+    if (!allocate_code_blob) {
+      return entry;
+    }
+  }
 #endif
 
-      new_adapter = AdapterBlob::create(&buffer);
-      NOT_PRODUCT(insts_size = buffer.insts_size());
-    }
-    if (new_adapter == NULL) {
-      // CodeCache is full, disable compilation
-      // Ought to log this but compile log is only per compile thread
-      // and we're some non descript Java thread.
-      return NULL; // Out of CodeCache space
-    }
-    entry->relocate(new_adapter->content_begin());
+  new_adapter = AdapterBlob::create(&buffer);
+  NOT_PRODUCT(int insts_size = buffer.insts_size());
+  if (new_adapter == NULL) {
+    fatal("No memory to create adapter");
```

Comment:
This used to be `return NULL;`, and the JVM is supposed to continue execution (but the linking of the current class will get a VirtualMachineError. I think we should keep the old behavior.

Reply:
Yes, an unintentional result of me breaking the code flow apart and then merging it back together.